### PR TITLE
Tyler's Stable SDK API implementation

### PR DESCRIPTION
**Pull request by cloutiertyler**
## Description of Changes

This implements the stable SDK API implementation

### CODE REVIEW

**Comment by PuruVJ** [comment]
One thing I had suggested in the proposal is using `AbortController` cancelling the events in my proposal, but I am also not against removeEvent functions. We should add AbortController to 1.0 release after this and emphasize using that over `remove` functions

File: packages/sdk/src/db_connection_builder.ts [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+   * @example
+   *
+   * ```ts
+   * spacetimeDBClient.onError((...args: any[]) => {
+   *  stdbLogger("warn","ERROR", args);
+   * });
+   * ```
+   */
+  onConnectError(
+    callback: (...args: any[]) => void
```
Comment (gefjon) [inline packages/sdk/src/db_connection_builder.ts:222]: Args should be `error: Error` or `error: any`.

File: packages/sdk/src/db_connection_builder.ts [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+   * Unlike callbacks registered via {@link DbConnection},
+   * no mechanism is provided to unregister the provided callback.
+   * This is a concession to ergonomics; there's no clean place to return a `CallbackId` from this method
+   * or from `build`.
+   *
+   * @param {function(error?: Error): void} callback - The callback to invoke upon disconnection.
+   * @throws {Error} Throws an error if called multiple times on the same `DbConnectionBuilder`.
+   */
+  onDisconnect(
+    callback: (...args: any[]) => void
```
Comment (gefjon) [inline packages/sdk/src/db_connection_builder.ts:255]: Args should be `conn: DbConnection, error?: Error | undefined`.

File: examples/quickstart/src/App.tsx [hunk 3]
```diff
@@ -1,25 +1,17 @@
...
 
-// Bindings
-import Message from './module_bindings/message';
-import SendMessageReducer from './module_bindings/send_message_reducer';
-import SetNameReducer from './module_bindings/set_name_reducer';
-import User from './module_bindings/user';
+import { DbConnection, type RemoteDBContext } from './module_bindings';
 
-import { Identity, SpacetimeDBClient } from '@clockworklabs/spacetimedb-sdk';
+import { DBConnectionBuilder, Identity } from '@clockworklabs/spacetimedb-sdk';
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/App.tsx:5]: Quickstart is still broken/not updated

File: examples/quickstart/src/App.tsx [hunk 4]
```diff
@@ -36,47 +28,57 @@ function App() {
...
   useEffect(() => {
-    if (!initialized.current) {
-      client.connect();
-      initialized.current = true;
+    async function main() {
+      if (!conn.current) {
+        conn.current = await DbConnection.builder()
+          .withUri('ws://localhost:3000')
+          .withModuleName('chat')
+          .withCredentials([Identity.fromString(identity!), token!])
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/App.tsx:39]: Quickstart is still broken/not updated

File: examples/quickstart/src/App.tsx [hunk 5]
```diff
@@ -36,47 +28,57 @@ function App() {
...
+          .onDisconnect(() => {
+            console.log('disconnected');
+          })
+          .onConnectError(() => {
+            console.log('client_error');
+          })
+          .onConnect((identity, token) => {
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/App.tsx:46]: Quickstart is still broken/not updated

File: examples/quickstart/src/App.tsx [hunk 6]
```diff
@@ -36,47 +28,57 @@ function App() {
...
-    client.on('client_error', () => {
-      console.log('client_error');
-    });
-
-    client.onConnect((token: string, identity: Identity) => {
-      console.log('Connected to SpacetimeDB');
+    if (!conn.current) return;
 
-      local_identity.current = identity;
+    // TODO: What do about this?
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/App.tsx:68]: Quickstart is still broken/not updated

File: examples/quickstart/src/module_bindings/index.ts [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+  init() {
+    this.connection.callReducer('__init__', new Uint8Array(0));
+  }
+
+  onInit(
+    callback: (ctx: EventContext<RemoteTables, RemoteReducers, Reducer>) => void
+  ) {
+    this.connection.onReducer('__init__', callback);
+  }
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/module_bindings/index.ts:143]: Ack

File: packages/sdk/src/db_connection_builder.ts [hunk 8]
```diff
@@ -0,0 +1,260 @@
+// TODO: This is temporary notes for how this should turn out. Will be removed by final PR
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply packages/sdk/src/db_connection_builder.ts:1]: Thank you

File: examples/quickstart/src/module_bindings/user_table.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  // Updates are only defined for tables with primary keys.
+  onUpdate = (
+    cb: (
+      ctx: EventContext<RemoteTables, RemoteReducers, Reducer>,
+      oldRow: User,
+      newRow: User
+    ) => void
+  ) => {
+    return this.tableCache.onUpdate(cb);
+  };
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/module_bindings/user_table.ts:133]: Ack

File: examples/quickstart/src/module_bindings/user.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
+// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/module_bindings/user.ts:1]: Ack

File: packages/sdk/src/db_connection_builder.ts (hunk 2 above)
Comment (cloutiertyler) [reply packages/sdk/src/db_connection_builder.ts:254]: Ack, thank you

File: packages/sdk/src/db_connection_builder.ts (hunk 1 above)
Comment (cloutiertyler) [reply packages/sdk/src/db_connection_builder.ts:221]: Ack, thank you

File: examples/quickstart/src/module_bindings/index.ts [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+  RemoteTables,
+  RemoteReducers
+> {
+  static builder = () => {
+    return new DBConnectionBuilder<DBConnection>(
+      REMOTE_MODULE,
+      (imp: DBConnectionImpl) => imp as DBConnection
+    );
+  };
+}
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/module_bindings/index.ts:208]: Done.

File: examples/quickstart/src/module_bindings/user.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/module_bindings/user.ts]: Done.

File: examples/quickstart/src/module_bindings/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/module_bindings/index.ts]: Done.

File: examples/quickstart/server/src/lib.rs [hunk 12]
```diff
@@ -1,55 +1,50 @@
 use anyhow::{anyhow, Result};
-use spacetimedb::{spacetimedb, Identity, ReducerContext, Timestamp};
+use spacetimedb::{Identity, ReducerContext, Table, Timestamp};
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/server/src/lib.rs:2]: I will investigate this. I think we'll need to develop a proper quickstart update process as well.

File: examples/quickstart/src/App.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/App.tsx:39]: Good question. Leaving unresolved for me to test

File: examples/quickstart/src/module_bindings/index.ts [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+// A type representing all the possible variants of a reducer.
+export type Reducer =
+  | { name: 'IdentityConnected'; args: IdentityConnected }
+  | { name: 'IdentityDisconnected'; args: IdentityDisconnected }
+  | { name: 'Init'; args: Init }
+  | { name: 'SendMessage'; args: SendMessage }
+  | { name: 'SetName'; args: SetName };
+
+export class RemoteReducers {
```
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/module_bindings/index.ts:122]: Done! I've implemented this.

File: examples/quickstart/src/module_bindings/user_table.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (cloutiertyler) [reply examples/quickstart/src/module_bindings/user_table.ts]: Done! I've implemented this.

File: packages/sdk/src/db_connection_builder.ts (hunk 1 above)
Comment (cloutiertyler) [reply packages/sdk/src/db_connection_builder.ts]: Done! I've implemented this.

File: packages/sdk/src/db_connection_builder.ts (hunk 2 above)
Comment (cloutiertyler) [reply packages/sdk/src/db_connection_builder.ts]: Done! I've implemented this.

**Pull request merged**